File: android/app/src/main/java/com/vocabreview/model/VocabWord.kt
package com.vocabreview.model

import kotlinx.serialization.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*
import kotlinx.serialization.json.*

typealias VocabMap = Map<String, VocabWord>
typealias MutableVocabMap = MutableMap<String, VocabWord>

@Serializable(with = VocabWordSerializer::class)
data class VocabWord(
    val status: String? = null,
    val firstSeen: String? = null,
    val lastSeen: String? = null,
    val contexts: List<VocabContext>? = null,
    val definition: String? = null,
    val phonetic: String? = null,
    val reviewCount: Int? = null,
    val correctCount: Int? = null,
    val missCount: Int? = null,
    val intervalDays: Int? = null,
    val dueAt: String? = null,
    val lastReviewed: String? = null,
    val additionalFields: Map<String, JsonElement> = emptyMap()
)

object VocabWordSerializer : KSerializer<VocabWord> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("VocabWord")

    private val knownKeys = setOf(
        "status", "firstSeen", "lastSeen", "contexts", "definition",
        "phonetic", "reviewCount", "correctCount", "missCount",
        "intervalDays", "dueAt", "lastReviewed"
    )

    override fun deserialize(decoder: Decoder): VocabWord {
        val jsonDecoder = decoder as JsonDecoder
        val obj = jsonDecoder.decodeJsonElement().jsonObject

        val contextsElement = obj["contexts"]
        val contexts: List<VocabContext>? = contextsElement?.let {
            if (it is JsonArray) {
                Json.decodeFromJsonElement<List<VocabContext>>(it)
            } else null
        }

        val additional = obj.filterKeys { it !in knownKeys }

        return VocabWord(
            status = obj["status"]?.jsonPrimitive?.contentOrNull,
            firstSeen = obj["firstSeen"]?.jsonPrimitive?.contentOrNull,
            lastSeen = obj["lastSeen"]?.jsonPrimitive?.contentOrNull,
            contexts = contexts,
            definition = obj["definition"]?.jsonPrimitive?.contentOrNull,
            phonetic = obj["phonetic"]?.jsonPrimitive?.contentOrNull,
            reviewCount = obj["reviewCount"]?.jsonPrimitive?.intOrNull,
            correctCount = obj["correctCount"]?.jsonPrimitive?.intOrNull,
            missCount = obj["missCount"]?.jsonPrimitive?.intOrNull,
            intervalDays = obj["intervalDays"]?.jsonPrimitive?.intOrNull,
            dueAt = obj["dueAt"]?.jsonPrimitive?.contentOrNull,
            lastReviewed = obj["lastReviewed"]?.jsonPrimitive?.contentOrNull,
            additionalFields = additional
        )
    }

    override fun serialize(encoder: Encoder, value: VocabWord) {
        val jsonEncoder = encoder as JsonEncoder
        val map = buildMap<String, JsonElement> {
            // Write additionalFields first (will be overwritten by known fields if conflict)
            putAll(value.additionalFields)
            // Write known fields
            value.status?.let { put("status", JsonPrimitive(it)) }
            value.firstSeen?.let { put("firstSeen", JsonPrimitive(it)) }
            value.lastSeen?.let { put("lastSeen", JsonPrimitive(it)) }
            value.contexts?.let { put("contexts", Json.encodeToJsonElement(it)) }
            value.definition?.let { put("definition", JsonPrimitive(it)) }
            value.phonetic?.let { put("phonetic", JsonPrimitive(it)) }
            value.reviewCount?.let { put("reviewCount", JsonPrimitive(it)) }
            value.correctCount?.let { put("correctCount", JsonPrimitive(it)) }
            value.missCount?.let { put("missCount", JsonPrimitive(it)) }
            value.intervalDays?.let { put("intervalDays", JsonPrimitive(it)) }
            value.dueAt?.let { put("dueAt", JsonPrimitive(it)) }
            value.lastReviewed?.let { put("lastReviewed", JsonPrimitive(it)) }
        }
        jsonEncoder.encodeJsonElement(JsonObject(map))
    }
}

File: ios/VocabReview/Models/VocabWord.swift
import Foundation

typealias VocabMap = [String: VocabWord]

struct VocabWord: Codable, Equatable {
    var status: String?
    var firstSeen: String?
    var lastSeen: String?
    var contexts: [VocabContext]?
    var definition: String?
    var phonetic: String?
    var reviewCount: Int?
    var correctCount: Int?
    var missCount: Int?
    var intervalDays: Int?
    var dueAt: String?
    var lastReviewed: String?
    var additionalFields: [String: JSONValue] = [:]

    var isLearning: Bool {
        status == "learning"
    }

    var latestSentence: String {
        contexts?.last?.sentence ?? ""
    }

    enum CodingKeys: String, CodingKey, CaseIterable {
        case status
        case firstSeen
        case lastSeen
        case contexts
        case definition
        case phonetic
        case reviewCount
        case correctCount
        case missCount
        case intervalDays
        case dueAt
        case lastReviewed
    }

    init(
        status: String? = nil,
        firstSeen: String? = nil,
        lastSeen: String? = nil,
        contexts: [VocabContext]? = nil,
        definition: String? = nil,
        phonetic: String? = nil,
        reviewCount: Int? = nil,
        correctCount: Int? = nil,
        missCount: Int? = nil,
        intervalDays: Int? = nil,
        dueAt: String? = nil,
        lastReviewed: String? = nil,
        additionalFields: [String: JSONValue] = [:]
    ) {
        self.status = status
        self.firstSeen = firstSeen
        self.lastSeen = lastSeen
        self.contexts = contexts
        self.definition = definition
        self.phonetic = phonetic
        self.reviewCount = reviewCount
        self.correctCount = correctCount
        self.missCount = missCount
        self.intervalDays = intervalDays
        self.dueAt = dueAt
        self.lastReviewed = lastReviewed
        self.additionalFields = additionalFields
    }

    init(from decoder: Decoder) throws {
        let known = try decoder.container(keyedBy: CodingKeys.self)
        status = try known.decodeIfPresent(String.self, forKey: .status)
        firstSeen = try known.decodeIfPresent(String.self, forKey: .firstSeen)
        lastSeen = try known.decodeIfPresent(String.self, forKey: .lastSeen)
        contexts = try known.decodeIfPresent([VocabContext].self, forKey: .contexts)
        definition = try known.decodeIfPresent(String.self, forKey: .definition)
        phonetic = try known.decodeIfPresent(String.self, forKey: .phonetic)
        reviewCount = try known.decodeIfPresent(Int.self, forKey: .reviewCount)
        correctCount = try known.decodeIfPresent(Int.self, forKey: .correctCount)
        missCount = try known.decodeIfPresent(Int.self, forKey: .missCount)
        intervalDays = try known.decodeIfPresent(Int.self, forKey: .intervalDays)
        dueAt = try known.decodeIfPresent(String.self, forKey: .dueAt)
        lastReviewed = try known.decodeIfPresent(String.self, forKey: .lastReviewed)

        let dynamic = try decoder.container(keyedBy: DynamicCodingKey.self)
        let knownKeys = Set(CodingKeys.allCases.map(\.rawValue))
        additionalFields = [:]

        for key in dynamic.allKeys where !knownKeys.contains(key.stringValue) {
            additionalFields[key.stringValue] = try dynamic.decode(JSONValue.self, forKey: key)
        }
    }

    func encode(to encoder: Encoder) throws {
        var dynamic = encoder.container(keyedBy: DynamicCodingKey.self)
        for (key, value) in additionalFields {
            try dynamic.encode(value, forKey: DynamicCodingKey(stringValue: key))
        }

        var known = encoder.container(keyedBy: CodingKeys.self)
        try known.encodeIfPresent(status, forKey: .status)
        try known.encodeIfPresent(firstSeen, forKey: .firstSeen)
        try known.encodeIfPresent(lastSeen, forKey: .lastSeen)
        try known.encodeIfPresent(contexts, forKey: .contexts)
        try known.encodeIfPresent(definition, forKey: .definition)
        try known.encodeIfPresent(phonetic, forKey: .phonetic)
        try known.encodeIfPresent(reviewCount, forKey: .reviewCount)
        try known.encodeIfPresent(correctCount, forKey: .correctCount)
        try known.encodeIfPresent(missCount, forKey: .missCount)
        try known.encodeIfPresent(intervalDays, forKey: .intervalDays)
        try known.encodeIfPresent(dueAt, forKey: .dueAt)
        try known.encodeIfPresent(lastReviewed, forKey: .lastReviewed)
    }
}

struct DynamicCodingKey: CodingKey {
    var stringValue: String
    var intValue: Int?

    init(stringValue: String) {
        self.stringValue = stringValue
    }

    init(intValue: Int) {
        self.stringValue = String(intValue)
        self.intValue = intValue
    }
}
